File: src/types/sensor.ts
// Telemetry and signal processing types for KanthaRakshak

export interface SensorDataPoint {
  timestamp: number; // relative ms from test start
  piezo: number; // normalized throat vibration amplitude (-1.0 to 1.0 or 0 to 1.0)
  ax: number; // acceleration X (g)
  ay: number; // acceleration Y (g)
  az: number; // acceleration Z (g)
  accel_magnitude: number; // resultant acceleration vector magnitude |a| (g)
}

export type SignalQualityLevel = 'GOOD' | 'FAIR' | 'POOR';
export type NoiseLevel = 'LOW' | 'MEDIUM' | 'HIGH';
export type MotionBaseline = 'STABLE' | 'MOVEMENT_DETECTED';
export type SensorAgreementLevel = 'HIGH' | 'MODERATE' | 'LOW' | 'DISAGREED';

export interface QualityPacket {
  type: 'quality';
  signal_quality: SignalQualityLevel;
  noise_level: NoiseLevel;
  motion_baseline: MotionBaseline;
  piezo_snr_db?: number;
}

export interface SwallowEventPacket {
  type: 'swallow_event';
  detected: boolean;
  timestamp?: number;
  confidence?: number;
  duration_ms?: number;
}

export interface SensorDataPacket extends SensorDataPoint {
  type: 'sensor_data';
}

export interface AnalysisCompletePacket {
  type: 'analysis_complete';
  test_id: string;
  result: 'LOW_RISK' | 'RETEST' | 'POSSIBLE_RISK';
}

export type WebSocketIncomingMessage =
  | SensorDataPacket
  | QualityPacket
  | SwallowEventPacket
  | AnalysisCompletePacket;

File: backend/app/schemas/sensor.py
from typing import Optional, Literal
from pydantic import BaseModel

class SensorSample(BaseModel):
    timestamp: float # relative ms
    piezo: float     # normalized -1.0 to 1.0 or voltage
    ax: float        # g
    ay: float        # g
    az: float        # g
    gx: float = 0.0  # deg/s
    gy: float = 0.0  # deg/s
    gz: float = 0.0  # deg/s
    accel_magnitude: float # resultant sqrt(ax^2 + ay^2 + az^2)

class QualityPacket(BaseModel):
    type: Literal["quality"] = "quality"
    signal_quality: Literal["GOOD", "FAIR", "POOR"]
    noise_level: Literal["LOW", "MEDIUM", "HIGH"]
    motion_baseline: Literal["STABLE", "MOVEMENT_DETECTED"]
    piezo_snr_db: Optional[float] = None

class SwallowEventPacket(BaseModel):
    type: Literal["swallow_event"] = "swallow_event"
    detected: bool
    timestamp: Optional[float] = None
    duration_ms: Optional[float] = None
    confidence: Optional[float] = None

class DeviceStatus(BaseModel):
    esp32_id: str
    ble_connected: bool
    ble_device_name: str
    firmware_version: str
    battery_percentage: int
    is_charging: bool
    packet_rate_hz: int
    last_packet_timestamp: Optional[str]
    piezo_adc_state: str
    mpu6050_state: str
    overall_status: str
